### PR TITLE
Annotate APIs in System.Private.Xml

### PR DESCRIPTION
**Pull request by agocke**
Mostly adds RequiresUnreferencedCode, and propagates the results
outward.

### CODE REVIEW

**Comment by eerhardt** [comment]
Can you remove any suppressions from https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.Xml/src/ILLink/ILLink.Suppressions.xml that are now marked with `RUC`?

File: src/libraries/System.Xml.ReaderWriter/ref/System.Xml.ReaderWriter.cs [hunk 1]
```diff
@@ -4,6 +4,8 @@
 // Changes to this file must follow the https://aka.ms/api-review process.
 // ------------------------------------------------------------------------------
 
+using System.Diagnostics.CodeAnalysis;
```
Comment (eerhardt) [inline src/libraries/System.Xml.ReaderWriter/ref/System.Xml.ReaderWriter.cs:7]: `ref` files are generated and use full qualified names in source. You shouldn't hand edit them because the next time someone wants to update it, they need to manually revert your changes.

Comment (eerhardt) [reply src/libraries/System.Xml.ReaderWriter/ref/System.Xml.ReaderWriter.cs:7]: See [https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/updating-ref-source.md](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/updating-ref-source.md) for how to update ref files.

---
In reply to: [595286512](https://github.com/dotnet/runtime/pull/49682#discussion_r595286512) [](ancestors = 595286512)

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchema.cs [hunk 2]
```diff
@@ -88,35 +89,41 @@ public class XmlSchema : XmlSchemaObject
...
             Write(xmlWriter, namespaceManager);
         }
 
+        [RequiresUnreferencedCode(XmlSerializer.LinkerSerializationMessage)]
         public void Write(XmlWriter writer)
         {
             Write(writer, null);
         }
 
+        [RequiresUnreferencedCode(XmlSerializer.LinkerSerializationMessage)]
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchema.cs:126]: Can't we make this safe by preserving the correct things on `XmlSchema`?

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/Mappings.cs [hunk 3]
```diff
@@ -1174,7 +1175,9 @@ internal string NamespaceList
 
         internal Type? Type
         {
+            [return: DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)]
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/Mappings.cs:1178]: Can just annotate the whole property instead of both the getter and setter.

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/Models.cs [hunk 4]
```diff
@@ -90,16 +94,20 @@ internal ArrayModel GetArrayModel(Type type)
     internal abstract class TypeModel
     {
         private readonly TypeDesc _typeDesc;
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)]
         private readonly Type _type;
         private readonly ModelScope _scope;
 
-        protected TypeModel(Type type, TypeDesc typeDesc, ModelScope scope)
+        protected TypeModel(
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)]
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/Models.cs:102]: (nit) this formatting is a little hard to read. I've usually been following the pattern of one param per line, and the attribute on the same line as the param it annoates:

```C#
protected TypeModel(
            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] Type type,
            TypeDesc typeDesc, 
            ModelScope scope)
```

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs [hunk 5]
```diff
@@ -17,17 +18,23 @@ namespace System.Xml.Serialization
 
     internal class ReflectionXmlSerializationReader : XmlSerializationReader
     {
+#pragma warning disable IL2026 // Methods annotated with 'RequiresUnreferencedCodeAttribute' require dynamic access otherwise can break functionality when trimming application code
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs:21]: This doesn't work for the linker since it can't see `#pragma`s. We will need to suppress them with `UnconditionalSuppressMessage` attributes.

As an aside - we know the types here, isn't there a way to make this safe?

Comment (eerhardt) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs:21]: (also -while we are in here, can we change these to be `readonly`? I don't think anyone sets them.)

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationWriter.cs [hunk 6]
```diff
@@ -14,8 +15,13 @@ internal class ReflectionXmlSerializationWriter : XmlSerializationWriter
     {
         private readonly XmlMapping _mapping;
 
+#pragma warning disable IL2026 // Methods annotated with 'RequiresUnreferencedCodeAttribute' require dynamic access otherwise can break functionality when trimming application code
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationWriter.cs:18]: Same comments as ReflectionXmlSerializationReader.cs

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/SchemaImporter.cs [hunk 7]
```diff
@@ -13,6 +13,7 @@
 
 #if DEBUG
     using System.Diagnostics;
+    using System.Diagnostics.CodeAnalysis;
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/SchemaImporter.cs:16]: You are using this outside of `#if DEBUG` below. Shouldn't this be in the group above?

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/SourceInfo.cs [hunk 8]
```diff
@@ -209,7 +215,9 @@ private void Convert(Type sourceType, Type? targetType, bool asAddress)
             }
         }
 
-        private void ConvertNullableValue(Type nullableType, Type targetType)
+        private void ConvertNullableValue(
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/SourceInfo.cs:219]: (nit) elsewhere you have `TrimmerConstants.AllMethods`. Can that be used here?

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/TypeExtensions.cs [hunk 9]
```diff
@@ -4,13 +4,16 @@
...
     using System.Reflection;
 
     internal static class TypeExtensions
     {
         private const string ImplicitCastOperatorName = "op_Implicit";
 
-        public static bool TryConvertTo(this Type targetType, object? data, out object? returnValue)
+        public static bool TryConvertTo(
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods
+                | DynamicallyAccessedMemberTypes.NonPublicMethods)] this Type targetType, object? data, out object? returnValue)
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/TypeExtensions.cs:16]: I just see `Public` used below. Do we really need `NonPublicMethods` on here?

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlAttributes.cs [hunk 10]
```diff
@@ -91,6 +93,7 @@ private static Type IgnoreAttribute
         /// <devdoc>
         ///    <para>[To be supplied.]</para>
         /// </devdoc>
+        [RequiresUnreferencedCode(XmlSerializer.LinkerSerializationMessage)]
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlAttributes.cs:96]: I'm not sure I see what requires unreferenced code in this constructor.

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/Types.cs [hunk 11]
```diff
@@ -1349,7 +1393,13 @@ internal static PropertyInfo GetDefaultIndexer(Type type, string? memberInfo)
             }
             return indexer;
         }
-        private static Type GetCollectionElementType(Type type, string? memberInfo)
+        private static Type GetCollectionElementType(
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/Types.cs:1397]: (nit) `TrimmerConstants.PublicMembers` ?

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlAttributes.cs [hunk 12]
```diff
@@ -74,6 +75,7 @@ internal XmlAttributeFlags XmlFlags
 
         private static Type IgnoreAttribute
         {
+            [RequiresUnreferencedCode("calls GetType")]
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlAttributes.cs:78]: Looking around, I can't find a `System.XmlIgnoreMemberAttribute` anywhere. Can the code that is looking for it be removed?

Comment (eerhardt) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlAttributes.cs:78]: Looking at [https://referencesource.microsoft.com/#mscorlib/system/xmlignorememberattribute.cs,33ba544d60bd65ba,references](https://referencesource.microsoft.com/#mscorlib/system/xmlignorememberattribute.cs,33ba544d60bd65ba,references), it appears it was only used for xml serializing AppDomainSetup. But that no longer appears to be the case in .NET Core.

---
In reply to: [595310314](https://github.com/dotnet/runtime/pull/49682#discussion_r595310314) [](ancestors = 595310314)

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchema.cs (hunk 2 above)
Comment (vitek-karas) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchema.cs:126]: I think it should be possible - but it would be pretty tricky. There's probably a lot of dependencies which would have to be added between this method and all the various properties and other magic. The XmlSchema object model is quite large...

**Comment by vitek-karas** [comment]
@LakshanF FYI

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchema.cs (hunk 2 above)
Comment (eerhardt) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchema.cs:126]: Isn't it better for us to do it, than have a customer who needs this in a trimmed app be forced to try to figure it out themselves?

Comment (agocke) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchema.cs:126]: I tried to stay away from intrusive code changes in this PR, just annotation. Happy to file a bug, though, if that's your preferred mechanism of tracking.

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/TypeExtensions.cs (hunk 9 above)
Comment (agocke) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/TypeExtensions.cs:16]: Hmm, the linker seems to think so, but I concur, this looks like a bug.

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlAttributes.cs (hunk 10 above)
Comment (agocke) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlAttributes.cs:96]: It calls the `IgnoreAttribute` property (which is RUC above)

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlAttributes.cs (hunk 12 above)
Comment (agocke) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlAttributes.cs:78]: Ah, and since it's fetched from corlib this should always return null, right?

**Comment by agocke** [comment]
@eerhardt Any way to automatically generate ILLink suppression files? Or is it totally manual?

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs [hunk 13]
```diff
@@ -150,29 +150,38 @@ private static XmlSerializerNamespaces DefaultNamespaces
             }
         }
 
+        // Linker warning messages
+        internal const string LinkerSerializationMessage = "Members from serialized types may be trimmed if not referenced directly";
```
Comment (MichalStrehovsky) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs:154]: Let's avoid the term "Linker" spreading outside the mono/linker repo.

The tool name is just plain unfortunate and we should keep it an implementation detail until we can fully get rid of it. It's already ambiguous in Xamarin (see e.g. here: https://github.com/xamarin/xamarin-macios/issues/10251#issuecomment-794368281 where it needs to disambiguate between "link native code" and "il link" instead of being able to use the standard computer term that everyone outside .NET understands for what it is - "link").

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchema.cs (hunk 2 above)
Comment (eerhardt) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchema.cs:126]: > intrusive code changes

I wasn't thinking about intrusive code changes. I was thinking we would fix this with adding the necessary `[DynamicDependency]` attributes, plus a trimming test or two to ensure it works.

**Comment by eerhardt** [comment]
> Any way to automatically generate ILLink suppression files? Or is it totally manual?

See https://github.com/dotnet/runtime/pull/44264/files, which is unmerged. You can pass `--generate-warning-suppressions xml` to the linker and it will generate the .xml file under the `artifacts\bin\ILLinkTrimAssembly` folder.

File: src/libraries/System.Data.Common/System.Data.Common.sln [hunk 14]
```diff
@@ -1,4 +1,8 @@
-﻿Microsoft Visual Studio Solution File, Format Version 12.00
+﻿
```
Comment (eerhardt) [inline src/libraries/System.Data.Common/System.Data.Common.sln:1]: @ViktorHofer - the .sln files are still being updated automatically. I assume we don't want to check in these changes, correct?

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlAttributes.cs (hunk 12 above)
Comment (eerhardt) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlAttributes.cs]: I've opened https://github.com/dotnet/runtime/issues/49761 for this. I'm not sure if we want to do this in your PR or separately. But, it looks like it will affect what you are marking as `RUC` here.

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs [hunk 15]
```diff
@@ -17,8 +18,18 @@ namespace System.Xml.Serialization
 
     internal class ReflectionXmlSerializationReader : XmlSerializationReader
     {
-        private static TypeDesc StringTypeDesc { get; set; } = (new TypeScope()).GetTypeDesc(typeof(string));
-        private static TypeDesc QnameTypeDesc { get; set; } = (new TypeScope()).GetTypeDesc(typeof(XmlQualifiedName));
+        private static readonly TypeDesc StringTypeDesc;
+        private static readonly TypeDesc QnameTypeDesc;
+
+        [RequiresUnreferencedCode("Calls GetTypeDesc")]
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs:24]: What does `RequiresUnreferencedCode` on a static ctor mean / do?

A different approach to suppress the warning would be to use the `Target` property, like in the following:

https://github.com/dotnet/runtime/blob/2d0fe49001423b18b520c9fae072c81e7a11c50c/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs#L18-L20

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/Types.cs [hunk 16]
```diff
@@ -1239,6 +1281,7 @@ private static TypeFlags GetConstructorFlags(Type type, ref Exception? exception
             return 0;
         }
 
+        [RequiresUnreferencedCode("creates XmlAttributes")]
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/Types.cs:1284]: Does this still need to be `RequiresUnreferencedCode` now that `XmlAttributes` is no longer marked as unsafe?

If it doesn't, i think a bunch of other stuff could be unmarked as well, like `ImportTypeDesc` and `GetTypeDesc`, which means other things, maybe even that static ctor issue could just go away.

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchema.cs [hunk 17]
```diff
@@ -88,35 +89,41 @@ public class XmlSchema : XmlSchemaObject
             return parser.XmlSchema;
         }
 
+        [RequiresUnreferencedCode(XmlSerializer.LinkerSerializationMessage)]
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchema.cs:92]: I think your changes to the corresponding `ref` file were lost in the latest commits.

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs (hunk 15 above)
Comment (agocke) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs:24]: It doesn't look like usage of the type with RUC on the static constructor warns today, but it seems like it should. Personally, I would expect it to work similarly to putting `ObsoleteAttribute` on a type -- every static usage of the type (including constructor) would warn, but instance invocations wouldn't.

Comment (MichalStrehovsky) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs:24]: > It doesn't look like usage of the type with RUC on the static constructor warns today, but it seems like it should.

I think Mateo was working on it recently. I have a separate conversation on whether my understanding of this will match the implementation, but I think for classes, marking the cctor as RUC should generate warnings on all constructors and static methods on the type that then need to be suppressed or propagated. The drop of illink in this repo probably doesn't have Mateo's changes yet or the annotation doesn't do what I think it needs to do (will find out soon).

Comment (vitek-karas) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs:24]: Even though it might work, the experience will be weird - it's basically equivalent to putting the RUC on the type (if we supported that). Is it really the case that the type is impossible to use safely with trimming - even for a little bit?

File: src/libraries/System.Data.Common/System.Data.Common.sln (hunk 14 above)
Comment (ViktorHofer) [reply src/libraries/System.Data.Common/System.Data.Common.sln:1]: Right. Please undo this change.

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/Types.cs (hunk 16 above)
Comment (agocke) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/Types.cs]: Unfortunately not that much benefit. Only GetAttributes seemed clean, and all the uses of that ended up needing RUC for some other reason.

Comment (eerhardt) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/Types.cs]: Oh well... it was a good try though.

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs (hunk 15 above)
Comment (eerhardt) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs:24]: > Is it really the case that the type is impossible to use safely with trimming - even for a little bit?

In this specific case, I believe we can safely suppress these warning since the hard-coded types used here `string` and `XmlQualifiedName` should be safe to pass to `GetTypeDesc`. Looking at the annotation for that method, it looks like it only gets into unsafe code if it calls `ImportTypeDesc`, which it should never do for these two types because they are considered "primitive" types:

https://github.com/dotnet/runtime/blob/f64246ce08fb7a58221b2b7c8e68f69c02522b0d/src/libraries/System.Private.Xml/src/System/Xml/Serialization/Types.cs#L493-L504

File: src/libraries/System.Data.Common/System.Data.Common.sln [hunk 18]
```diff
@@ -38,36 +42,27 @@ EndProject
...
-		{09129981-6B0D-44C7-9F0A-AD2426106332} = {2FBD1B5D-0E5B-4D8C-A228-48B428725D67}
-		{0BD22D2A-C12C-4641-8F12-73D21AAAFBA3} = {12B0E17A-6957-425E-98E1-5F245010893A}
-		{BEBD7B5B-9544-42EB-B878-F009560CAAF4} = {12B0E17A-6957-425E-98E1-5F245010893A}
-		{5305244F-C52B-4772-A13D-C034213F3833} = {9A105AFB-AA36-46E1-BC68-6724F5644D0C}
-		{F865C526-79F8-403B-B5E4-52A90ECE21E3} = {9A105AFB-AA36-46E1-BC68-6724F5644D0C}
-		{A404332F-CCAA-4E22-94C1-635F7F19176F} = {9A105AFB-AA36-46E1-BC68-6724F5644D0C}
-		{BF66C27C-D1B1-4F86-A1E1-EDC63584E818} = {9A105AFB-AA36-46E1-BC68-6724F5644D0C}
-		{649CA543-288D-4A8B-AED3-E470DC7FF866} = {9A105AFB-AA36-46E1-BC68-6724F5644D0C}
+	GlobalSection(SharedMSBuildProjectFiles) = preSolution
+		..\System.Private.CoreLib\src\System.Private.CoreLib.Shared.projitems*{f3dc1df9-698a-4970-9a5e-ab946d6f763b}*SharedItemsImports = 5
```
Comment (joperezr) [inline src/libraries/System.Data.Common/System.Data.Common.sln:46]: Can you undo the changes in the sln file? These are generated by a script, so it is better if we try to keep them just updated with that script.

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlAttributes.cs [hunk 19]
```diff
@@ -3,6 +3,7 @@
 
 using System.Reflection;
 using System.ComponentModel;
+using System.Diagnostics.CodeAnalysis;
```
Comment (joperezr) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlAttributes.cs:6]: NIT: seems like this is not really needed.

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlAttributes.cs [hunk 20]
```diff
@@ -99,7 +83,7 @@ public XmlAttributes(ICustomAttributeProvider provider)
             XmlAnyElementAttribute? wildcard = null;
             for (int i = 0; i < attrs.Length; i++)
             {
-                if (attrs[i] is XmlIgnoreAttribute || attrs[i] is ObsoleteAttribute || attrs[i].GetType() == IgnoreAttribute)
+                if (attrs[i] is XmlIgnoreAttribute || attrs[i] is ObsoleteAttribute)
```
Comment (joperezr) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlAttributes.cs:86]: Not sure why we used to have that IgnoreAttribute using reflection if we already had the check for XmlIgnoreAttribute here. Do you know why we had it in both places?

Comment (eerhardt) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlAttributes.cs:86]: It was originally to break a circular dependency between `mscorlib` (where AppDomainSetup lived) and `System` (where Xml serialization and `XmlIgnoreAttribute` lived). So they could ignore certain members on AppDomainSetup when it was serialized.

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchema.cs (hunk 2 above)
Comment (joperezr) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchema.cs]: I've addressed this by keeping all public members on XmlSchema and validated with a trimming test that I'm adding as well.

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/TypeExtensions.cs (hunk 9 above)
Comment (joperezr) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/TypeExtensions.cs]: I removed the nonPublicMethods and the linker was no longer warning so if it was a bug it looks like it is fixed now.

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs (hunk 13 above)
Comment (joperezr) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs]: removed.

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchema.cs [hunk 21]
```diff
@@ -117,6 +118,10 @@ public void Write(XmlWriter writer)
             Write(writer, null);
         }
 
+        [DynamicDependency(TrimmerConstants.PublicMembers, typeof(XmlSchema))]
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "Supressing warning since the right members of XmlSchema will be preserved by the " +
+            "DynamicDependency attribute.")]
```
Comment (vitek-karas) [inline src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchema.cs:124]: This is not enough unfortunately. Serialization is recursive - the dynamic dependency you added will only work on `XmlSchema` itself, not any of the other types the serializer will get to by recursively serialization data on the schema. @sbomer and/or @LakshanF are working on adding some support for `XmlSerializer` based on custom steps. I *think* it could make this work if fully implemented (the `XmlSchema` has `XmlRootAttribute` on it, so it's discoverable by the custom step). I don't know how far we've got though.

Comment (vitek-karas) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchema.cs:124]: All that said I would probably look at prioritizing this. I don't think it's very common to write `xsd` files in general - and in trimmed app on top...

Comment (joperezr) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchema.cs:124]: I see, I'm not very familiar with XmlSchema but do you have an example of a type of Schema that would not work with just this DynamicDependency so that I can add a trimming test for it and to figure out what else needs to be preserved? I made sure that the trimming test I added in the last commit was failing without the DynamicDependency attribute and it is passing now after adding that annotation.

Comment (vitek-karas) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchema.cs:124]: I'd have to play with it - but high level:
* The XmlSchema creates an OM representation of the schema in memory
* The OM is basically a graph of XmlSchemaObject and derived classes (plus probably some other helper classes and so on)
* The serialization will walk that OM and write it out into XML - in great detail (after all it should be an exact serialization - meaning you should be able to roundtrip the OM through the file).
* That means the serialization will need pretty the entire OM preserved

Basically I would have to go and find a hidden property somewhere for which nobody calls a getter - create the object it belongs to and pass it to the serialization -> boom.
I guess what I'm trying to say - this can be pretty costly to get right - we would need to somehow figure out (and ideally prove) the full graph the serializer can walk in this case.

Comment (eerhardt) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchema.cs:124]: I'd be fine with adding the test, `DynamicDependency`, and `RequiresUnreferencedCode` for now. Users will still get a warning if they are calling this to let them know it may not work. We can always remove the `RequiresUnreferencedCode` in the future once we are 100% certain everything will work correctly.

Comment (vitek-karas) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchema.cs:124]: I like @eerhardt suggestion.

File: src/libraries/System.Private.Xml/tests/TrimmingTests/XmlSchema.Write.cs [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+
+        // <xs:element name="pets">
+        XmlSchemaElement elementPets = new XmlSchemaElement();
+        schema.Items.Add(elementPets);
+        elementPets.Name = "pets";
+
+        using (var stream = new MemoryStream())
+        using (var writer = XmlWriter.Create(stream))
+        {
+            schema.Write(writer, null);
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/tests/TrimmingTests/XmlSchema.Write.cs:45]: Can we do some simple verification this is correct? maybe we have a hard-coded string that we compare against?

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlReflectionImporter.cs [hunk 23]
```diff
@@ -1291,6 +1321,7 @@ private MembersMapping ImportMembersMapping(XmlReflectionMember[] xmlReflectionM
             return members;
         }
 
+        [RequiresUnreferencedCode("calls .ctor")]
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlReflectionImporter.cs:1324]: This one is odd - calls what ctor?

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlReflectionImporter.cs [hunk 24]
```diff
@@ -2219,6 +2254,7 @@ private static ElementAccessor CreateElementAccessor(TypeMapping mapping, string
         }
 
         // will create a shallow type mapping for a top-level type
+        [RequiresUnreferencedCode("calls .ctor")]
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlReflectionImporter.cs:2257]: Same as above - which ctor?

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs [hunk 25]
```diff
@@ -399,17 +416,19 @@ private object GenerateEncodedMembersElement(XmlMembersMapping xmlMembersMapping
...
-                          if (Reader.GetAttribute("id", null) != null)
-                          {
-                              ReadReferencedElement();
-                          }
-                          else
-                          {
-                              UnknownNode(p);
-                          };
-                      };
+                    unrecognizedElementSource = Wrapper;
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs:419]: @agocke - is there a performance penalty for this pattern? This issue is still open - [https://github.com/dotnet/roslyn/issues/5835](https://github.com/dotnet/roslyn/issues/5835).

Or is the idea that since the lambda method already capturing `this`, there is no perf difference?

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlReflectionImporter.cs (hunk 23 above)
Comment (joperezr) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlReflectionImporter.cs]: For both places, I've changed the comment to point to the right method that causes this one to be unsafe.

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs [hunk 26]
```diff
@@ -150,29 +150,38 @@ private static XmlSerializerNamespaces DefaultNamespaces
             }
         }
 
+        // Trimmer warning messages
+        internal const string TrimSerializationWarning = "Members from serialized types may be trimmed if not referenced directly";
+        private const string s_trimDeserializationWarning = "Members from deserialized types may be trimmed if not referenced directly";
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs:155]: (nit) since this isn't static (it is const), it should be named `TrimDeserializationWarning`

File: src/libraries/System.Private.Xml/tests/TrimmingTests/XmlSchema.Write.cs [hunk 27]
```diff
@@ -0,0 +1,62 @@
...
+        elementPets.Name = "pets";
+
+        using (var stream = new MemoryStream())
+        {
+            using (var writer = XmlWriter.Create(stream))
+            {
+                schema.Write(writer);
+            }
+
+            var str = Encoding.ASCII.GetString(stream.ToArray());
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/tests/TrimmingTests/XmlSchema.Write.cs:54]: (minor) Should be `UTF8`. Not that it really matters, but technically that's what it should be.

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs (hunk 26 above)
Comment (agocke) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs:155]: (nit: `const`s are static members according to the C# language specification 😄)

Comment (eerhardt) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs:155]: Tell that to our https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md

> 3. We use `_camelCase` for internal and private fields and use `readonly` where possible. Prefix internal and private instance fields with `_`, static fields with `s_`

> 12. We use PascalCasing to name all our constant local variables and fields.

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs (hunk 25 above)
Comment (agocke) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs:419]: Technically local functions are not method groups. This should produce the exact same IL as a lambda with the same capture semantics.

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationWriter.cs [hunk 28]
```diff
@@ -12,10 +13,20 @@ namespace System.Xml.Serialization
...
     internal sealed class ReflectionXmlSerializationWriter : XmlSerializationWriter
     {
-        private readonly XmlMapping _mapping;
+        internal static readonly TypeDesc StringTypeDesc;
+        internal static readonly TypeDesc QnameTypeDesc;
+
+        [RequiresUnreferencedCode("class GetTypeDesc")]
+        // Initialize in static constructor to suppress RequiresUnreferencedCode warning
+#pragma warning disable CA1810
+        static ReflectionXmlSerializationWriter()
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationWriter.cs:22]: Can you fix this one too? Or maybe even better refactor it to use the one from the reader so it is only in 1 spot??

**Review by eerhardt — APPROVED** [review]
Just one last comment to fix the Writer like we did for the Reader. Fix that and then we should merge this in.

**Comment by agocke** [comment]
Thanks everyone!
